### ISSUE TITLE
"? super V" leads to unsatisfiable bounds

### ISSUE THREAD

**Issue by jameslyon** [typetools/checker-framework#824]
The interpretation of "? super V" is "? super V extends @NonNull Object". In the Guava Futures library, there are functions such as:

```java
static <I extends @Nullable Object, O extends @Nullable Object> ListenableFuture<O> transform(ListenableFuture<I> input, Function<? super I,? extends O> function);
```

This leads to the following code failing to type check:

```java
public static ListenableFuture</*@NullableType*/ String> nullableStringTransform(
    ListenableFuture</*@NullableType*/ String> input,
    Function</*@NullableType*/ String, /*@NullableType*/ String> f) {
  return Futures.transform(input, f);
}
```

This appears to be related to us defining annotations in astub files rather than main source code, it doesn't seem to be reproducible if I define equivalent interfaces inline. The error reported is:

```
  found   : @Initialized @NonNull Function<@Initialized @Nullable String, @Initialized @Nullable String>
  required: @Initialized @NonNull Function<? extends @Initialized @NonNull Object super @Initialized @Nullable String, ? extends @Initialized @Nullable String>
```

There is also a second problem. If I change the astub file by adding `@Nullable` to the `Function` parameter like so:

```java
static <I extends @Nullable Object, O extends @Nullable Object> ListenableFuture<O> transform(ListenableFuture<I> input, Function<@Nullable ? super I,? extends O> function);
```

I then cannot compile a non-null version of the method above:

```java
public static ListenableFuture<String> stringTransform(
    ListenableFuture<String> input,
    Function<String, String> f) {
  return Futures.transform(input, f);
}
```

That fails with the error:

```
  found   : @Initialized @NonNull Function<@Initialized @NonNull String, @Initialized @NonNull String>
  required: @Initialized @NonNull Function<@Nullable ? extends @Initialized @Nullable Object super @Initialized @Nullable String, ? extends @Initialized @NonNull String>
```

This seems to contradict the checker framework documentation which states that in `@Nullable X super T` `@Nullable` is an upper rather than a lower bound, and therefore should be substitutable with a non-null type.

**Comment by mernst** [comment]
Can you please produce a reproducible test case and details about what went wrong?
You have provided some snippets of your code, and some snippets of the output, but we need complete information.  (We've tried many things already in an attempt to reproduce your test case, but we just ended up wasting time.)

There is a section in the manual about how to report a bug: http://types.cs.washington.edu/checker-framework/current/checker-framework-manual.html#reporting-bugs

Thanks for helping us help you.  And thanks for letting us know about this problem you encountered.

**Comment by mernst** [comment]
It has been a week without any response, so I'm closing the issue.
If any can reproduce the problem, please reopen the issue.  Thanks!

**Comment by jameslyon** [comment]
Sorry for the slow reply. I've been away, and it took me a while to reproduce in a vanilla build environment. You need three files:

First.java
```java
public class First {
  public interface Supplier<T> {}
  public interface Callable<T> {}
  public static <T> void method(Supplier<T> supplier, Callable<? super T> callable) {}
}
```

Second.java
```java
/*>>>import org.checkerframework.checker.nullness.qual.Nullable;*/

public class Second {
  public static void one(First.Supplier<Integer> supplier, First.Callable</*@Nullable*/ Object> callable) {
    First.method(supplier, callable);
  }

  public static void two(First.Supplier<Integer> supplier, First.Callable<Object> callable) {
    First.method(supplier, callable);
  }
}
```

First.astub
```
import org.checkerframework.checker.nullness.qual.Nullable;

public class First {
  public static <T extends @Nullable Object> void method(Supplier<T> supplier, Callable<? super T> callable) {}
}

public interface First$Supplier<T extends @Nullable Object> {}

public interface First$Callable<T extends @Nullable Object> {}
```

Compile First.java with then OpenJDK javac, then compile Second.java using the checker framework javac with the command

```
javac -processor org.checkerframework.checker.nullness.NullnessChecker -Astubs=First.astub -cp . Second.java
```

That yields:

```
Second.java:5: error: [argument.type.incompatible] incompatible types in argument.
    First.method(supplier, callable);
                           ^
  found   : @Initialized @NonNull Callable<@Initialized @Nullable Object>
  required: @Initialized @NonNull Callable<? extends @Initialized @NonNull Object super @Initialized @NonNull Integer>
```

Changing the astub file to read `Callable<@Nullable ? super T>` instead of `Callable<? super T>` yields a different error in method two:

```
Second.java:9: error: [argument.type.incompatible] incompatible types in argument.
    First.method(supplier, callable);
                           ^
  found   : @Initialized @NonNull Callable<@Initialized @NonNull Object>
  required: @Initialized @NonNull Callable<@Nullable ? extends @Initialized @Nullable Object super @Initialized @Nullable Integer>
1 error
```

This is probably two separate bugs. I'd argue that `? super T` in the astub file should match nullable types. The second issue is that according to my understanding from the manual `@Nullable ? super T` should mean that the upper bound is nullable, and therefore be equivalent to the pseudocode `? super T extends @Nullable object`, but that doesn't seem to be the case.

**Comment by smillst** [comment]
@jameslyon Thanks for the test case.  I am able to reproduce this bug.  I have added a test case checker/jtreg/nullness/issue824.
